### DISCUSSION TITLE
Improve PICKUP Mode to Allow Modulation Recording from External MIDI Controllers

### DISCUSSION THREAD

**Discussion by ghost** [SynthstromAudible/DelugeFirmware#2919] [Ideas]
Currently, when using the **PICKUP** mode in **MIDI > TAKEOVER**, the Deluge ignores changes to its internal parameters until the external MIDI controller’s encoder or fader matches the Deluge’s current parameter value. While this behavior is useful for avoiding sudden jumps in parameter values, it prevents recording smooth modulations or automations directly from an external controller.

This issue affects scenarios where users want to use MIDI controllers (e.g., Faderfox PC12 or LaunchControl XL) to record dynamic automations. In these cases, the Deluge does not respond to input until the PICKUP condition is met, resulting in a lack of recorded data or broken automation curves.

**Steps to Reproduce:**

Set **MIDI > TAKEOVER > PICKUP** mode on the Deluge.
Map an external MIDI encoder or fader to a parameter (e.g., filter cutoff).
Attempt to record an automation by moving the external control while the parameter values are not yet "picked up."
Observe that no automation data is recorded until the PICKUP threshold is matched, causing erratic or incomplete results.
Expected Behavior:
In PICKUP mode, the Deluge should still allow automation recording even if the external controller hasn’t matched the internal parameter value yet. The recorded automation should reflect the values sent by the MIDI controller in real time.

**Proposed Solution:**

Allow automation recording to bypass the PICKUP restriction while retaining PICKUP behavior for live performance and value jumps.
Alternatively, add a new mode or toggle that temporarily switches to JUMP behavior during recording, ensuring smooth modulation capture.
Additional Context:
This issue was tested with multiple MIDI controllers (e.g., Faderfox PC12, LaunchControl XL) and occurs consistently in the latest firmware (Beta 1.2 - Chopin). Automation recording works perfectly in JUMP mode but is impractical in PICKUP mode.

This improvement would significantly enhance the workflow for users relying on external controllers for modulation recording while keeping PICKUP's benefits in other contexts.

**Comment by seangoodvibes** [comment]
I think we could accommodate this and check if recording is enabled when processing incoming CC's. If enabled, go to the jump branch (if takeover mode is not relative) or the relative cc branch if otherwise

**Reply by ghost** [reply]
If this feature can be implemented, it would be fantastic for workflows involving external controllers. In the meantime, I’ll happily stick to using the Golden Knob for recording my modulations, which is already an excellent tool.

Big thanks to the team and contributors for pushing the Deluge even further with these thoughtful updates!

**Comment by duncanrmi** [comment]
I'm a bit puzzled- when you're recording new automation data in this way, what do you expect the automated parameter to do- what do you expect to hear?- if the incoming data hasn't matched the current parameter value yet? when will the pickup happen? there's going to have to be a point where the automation data matches the current value, or else there's going to be a jump when the recorded data suddenly becomes 'live'.... isn't there?

**Reply by seangoodvibes** [reply]
When recording midi takeover modes pickup and scale will behave like jump

I have just done this and opened a PR if you want to try it

I also made it behave this way when you're step editing (eg holding a step)

**Reply by duncanrmi** [reply]
ah, so when you drop into recording automation, you straightaway jump to the live value? that makes sense....

**Reply by seangoodvibes** [reply]
> ah, so when you drop into recording automation, you straightaway jump to the live value? that makes sense.... 

Yea for recording automation it will jump to the controller value immediately.

Then when you stop recording it will go back to the normal takeover mode behaviour.

**Comment by seangoodvibes** [comment]
@wiliamailysh i fixed this here: https://github.com/SynthstromAudible/DelugeFirmware/pull/2989
